### PR TITLE
Implement support for (most of) border-image.

### PR DESCRIPTION
**Pull request by glennw**
Add support for borders that use:
 * border-image-source (images only, not gradients yet).
 * border-image-slice
 * border-image-repeat (repeat + stretch only, for now).
 * border-image-outset

The major part of the work is supporting sub-rects when drawing images.

The rest of the work is simply generating image primitives from the
border-image display item description.

---
This change is [Reviewable](https://reviewable.io/reviews/servo/webrender/895)

### CODE REVIEW

**Comment by glennw** [comment]
r? @kvark 

cc @jrmuizel 

This will also need a version number bump once all comments are addressed, before it can be landed.

The corresponding Servo patch is here if you want to do some testing with it: https://github.com/glennw/servo/tree/border-image

**Comment by jrmuizel** [comment]
I had a quick look at this and it looks good to me.

File: webrender/src/frame_builder.rs [hunk 1]
```diff
@@ -254,118 +299,223 @@ impl FrameBuilder {
...
-                                   LayerSize::new(right.width,
-                                                  rect.size.height - top.width - bottom.width)),
-                    LayerRect::new(LayerPoint::new(bl_outer.x, bl_inner.y),
-                                   LayerSize::new(rect.size.width, bottom.width))
+                      border_item: &BorderDisplayItem) {
+        match border_item.details {
+            BorderDetails::Image(ref border) => {
+                // Calculate the modified rect as specific by border-image-outset
+                let origin = LayerPoint::new(rect.origin.x - border.outset.left,
+                                             rect.origin.x - border.outset.top);
```
Comment (kvark) [inline webrender/src/frame_builder.rs]: was this meant to be `rect.origin.y - border.outset.top`?

File: webrender/src/frame_builder.rs [hunk 2]
```diff
@@ -25,13 +25,58 @@ use tiling::{PrimitiveFlags, PrimitiveRunCmd, RenderPass, RenderTargetContext};
...
+}
+
+impl ImageBorderSegment {
+    fn new(rect: LayerRect,
+           sub_rect: TexelRect,
+           repeat_horizontal: RepeatMode,
+           repeat_vertical: RepeatMode) -> ImageBorderSegment {
+        let tile_spacing = LayerSize::zero();
+
+        let image_size = LayerSize::new(sub_rect.uv1.x - sub_rect.uv0.x,
```
Comment (kvark) [inline webrender/src/frame_builder.rs]: does it handle `uv1 < uv0` case correctly?

**Comment by glennw** [comment]
@kvark Thanks! Fixed those issues and rebased. With the corresponding Servo PR, this adds 13 new test passes and no new failures.

**Comment by glennw** [comment]
@bors-servo r=kvark

(Fixed a tidy error).

**Pull request merged**